### DISCUSSION TITLE
Keycloak custom SPI advice

### DISCUSSION THREAD

**Discussion by hillout** [keycloak/keycloak#21510] [Q&A]
Hello everyone,

Texting here cuz I'm trying to figure out what approach I need to follow in order to achieve business goal. So, here is the deal - in our company we have a standard user federation ldap/ad, and authentication to some internal systems goes through Keycloak - Kerberos - AD, works great. There is also a standalone Database user storage, where basically stores a lot of user attribute information (dynamically changes over time) in a format key = value, and this information is used by many other internal systems for authorization which aren't tied with Keycloak yet. So, in order to do so, meaning those systems will work with Keycloak, instead of System goes to the Database for the user attributes, Keycloak goes to that Database and once user gets authorized via Keycloak/Kerberos/AD chain in the token the system gets an additional information from that Database. On my opinion it has to be like either of these options: 

1. A User Storage SPI written from 0 which as a master information uses that Database (but there are no credentials there only attributes), and as it turned out it is not a solution cuz in the future this database will be deprecated. 
2. To re-wright the existing User Storage SPI LDAP, so it'll work in a way - once the authentication is done it somehow goes to the database, finds the user and maps attributes needed. 

Please share any thoughts, am I completely wrong and should I take another approach or something? Or maybe I'm thinking in the right direction.. 

Thanks in advance.

**Comment by christian-2** [comment]
Could you not sync those attributes from your database into your LDAP server somehow? Then unmodified Keycloak could pick them up via  [user attributes](https://www.keycloak.org/docs/latest/server_admin/index.html#proc-configuring-user-attributes_server_administration_guide) and [LDAP mappers](https://www.keycloak.org/docs/latest/server_admin/index.html#_ldap_mappers).

**Reply by hillout** [reply]
That's the first thing I was thinking of, but it's not the solution. Historically all the attributes for users store in that base and keep filling over time.

**Comment by hillout** [comment]
Implemented own SPI.
